cpp/src/parquet/encryption_internal.cc

```diff
@@ -0,0 +1,613 @@
+  memset(nonce, 0, kNonceLength);
+  // Random nonce
+  RAND_bytes(nonce, sizeof(nonce));
+
+  if (kGcmMode == aes_mode_) {
+    return GcmEncrypt(plaintext, plaintext_len, key, key_len, nonce, aad, aad_len,
+                      ciphertext);
+  }
+
+  return CtrEncrypt(plaintext, plaintext_len, key, key_len, nonce, ciphertext);
+}
+
+int AesEncryptor::AesEncryptorImpl::GcmEncrypt(const uint8_t* plaintext,
+                                               int plaintext_len, uint8_t* key,
+                                               int key_len, uint8_t* nonce, uint8_t* aad,
+                                               int aad_len, uint8_t* ciphertext) {
+  int len;
+  int ciphertext_len;
+
+  uint8_t tag[kGcmTagLength];
+  memset(tag, 0, kGcmTagLength);
+
+  // Setting key and IV (nonce)
+  if (1 != EVP_EncryptInit_ex(ctx_, nullptr, nullptr, key, nonce)) {
+    throw ParquetException("Couldn't set key and nonce");
+  }
+
+  // Setting additional authenticated data
+  if ((nullptr != aad) && (1 != EVP_EncryptUpdate(ctx_, nullptr, &len, aad, aad_len))) {
+    throw ParquetException("Couldn't set AAD");
+  }
+
+  // Encryption
+  if (1 != EVP_EncryptUpdate(ctx_, ciphertext + kBufferSizeLength + kNonceLength, &len,
+                             plaintext, plaintext_len)) {
+    throw ParquetException("Failed encryption update");
+  }
+
+  ciphertext_len = len;
+
+  // Finalization
+  if (1 != EVP_EncryptFinal_ex(ctx_, ciphertext + kBufferSizeLength + kNonceLength + len,
+                               &len)) {
+    throw ParquetException("Failed encryption finalization");
+  }
+
+  ciphertext_len += len;
+
+  // Getting the tag
+  if (1 != EVP_CIPHER_CTX_ctrl(ctx_, EVP_CTRL_GCM_GET_TAG, kGcmTagLength, tag)) {
+    throw ParquetException("Couldn't get AES-GCM tag");
+  }
+
+  // Copying the buffer size, nonce and tag to ciphertext
+  int buffer_size = kNonceLength + ciphertext_len + kGcmTagLength;
+  ciphertext[3] = static_cast<uint8_t>(0xff & (buffer_size >> 24));
+  ciphertext[2] = static_cast<uint8_t>(0xff & (buffer_size >> 16));
+  ciphertext[1] = static_cast<uint8_t>(0xff & (buffer_size >> 8));
+  ciphertext[0] = static_cast<uint8_t>(0xff & (buffer_size));
+  std::copy(nonce, nonce + kNonceLength, ciphertext + kBufferSizeLength);
+  std::copy(tag, tag + kGcmTagLength,
+            ciphertext + kBufferSizeLength + kNonceLength + ciphertext_len);
+
+  return kBufferSizeLength + buffer_size;
+}
+
+int AesEncryptor::AesEncryptorImpl::CtrEncrypt(const uint8_t* plaintext,
+                                               int plaintext_len, uint8_t* key,
+                                               int key_len, uint8_t* nonce,
+                                               uint8_t* ciphertext) {
+  int len;
+  int ciphertext_len;
+
+  // Parquet CTR IVs are comprised of a 12-byte nonce and a 4-byte initial
+  // counter field.
+  // The first 31 bits of the initial counter field are set to 0, the last bit
+  // is set
+  // to 1.
```

Comment:
is it necessary to break the line twice?

Reply:
I missed that. will fix it. Thanks